### DISCUSSION TITLE
Component pour Home Assistant

### DISCUSSION THREAD

**Discussion by webjib** [bokub/conso-api#4] [Aide] [answered]
Malgré quelques recherches, je n’ai pas trouvé de component pour Home Assistant, qui permettrait de faire appel à l’API et d’exposer les données au sein de Home Assistant. Quelle serait la meilleure option ?

**Comment by bokub** [comment]
Salut,

Je n'ai pour l'instant pas trouvé de moyen "simple" de faire ça, la meilleure option serait donc qu'un développeur ayant les compétences crée un add-on ou une intégration Home Assistant (ce n'est pas mon cas)

Si tu es membre du forum [HACF](https://forum.hacf.fr/), je t'invite à y suggérer l'idée !

**Reply by webjib** [reply]
Merci pour ta réponse ! Je vais essayer de voir si j’arrive à développer quelque chose, même si le python n’est pas ce que je préfère 😅

**Reply by bokub** [reply]
N'hésite surtout pas à demander si tu as des questions ou besoin de suggestions !

**Reply by webjib** [reply]
Merci ! J'ai commencé à bricoler un truc, tu pourras identifier mon intégration avec "mylinky-hass/0.0.1" en User-Agent. J'essaie d'économiser au max les API calls dans ma phase de développement ;-) Dès que j'ai un truc pas trop bancal, je mettrais ça sur GitHub.

**Reply by webjib** [reply]
Hello, sais tu s’il est envisageable de mettre en place un endpoint permettant de récupérer tous les types de données en une seul requête ? Cela permettrait d’économiser des API calls, et ça faciliterait le travail pour ma nouvelle intégration. Si ce n’est pas possible, et que l’utilisateur souhaite récupérer les 5 types de données, ça veut dire que je dois lancer en parallèle 5 requêtes :-/

**Reply by bokub** [reply]
Hello !
Alors non tu vas devoir faire des requêtes séparément, mais plusieurs choses à savoir :
- Les données de consommation et de production viennent forcément de PRM différents donc ce sera maximum 3 requêtes en parallèle pour un PRM donné
- Si tu récupères la courbe de charge, ça ne sert à rien de récupérer la consommation/production quotidienne pour le même jour, il suffit d'additionner les consommations de chaque demi-heure il me semble

**Comment by webjib** [comment]
Je viens de faire un test sur une journée, mais en additionnant les données par demi-heure, je ne retombe pas du tout sur la consommation totale de la journée. Y a t il une subtilité qui m'échappe ?

**Reply by bokub** [reply]
Aucune idée 🤷

**Reply by bokub** [reply]
Hello, je viens de faire le test et effectivement je n'obtiens pas le même résultat mais exactement le double !

L'explication est très simple: sur la courbe de charge, les valeurs correspondent à la **puissance** moyenne sur l'intervalle, alors que sur la journée, il s'agit de l'**énergie** consommée sur la journée.

Prenons un exemple concret, si je considère les hypothèses suivantes:
- ma puissance moyenne de 00:00 à 00:30 est de 100W
- ma puissance moyenne de 00:30 à 01:00 est de 200W

Alors ma puissance **moyenne** entre 00:00 et 01:00 est de 150W. Ma consommation sur ce créneau s'élève donc à 150Wh (150W pendant 1h), et non pas 300W (la somme de 100 et 200).

Pour aller encore plus dans le détail, il faudrait idéalement multiplier chaque puissance moyenne par la portion d'heure pour la transformer en énergie

Par exemple, avec des intervalles de 30 minutes:
- ma puissance moyenne de 00:00 à 00:30 est de 100W ->`0.5 * 100 = 50 Wh consommés`

Et avec des intervalles de 10 minutes (c'est possible parfois):
- ma puissance moyenne de 00:00 à 00:10 est de 100W ->`(10 / 60) * 100 ~= 16.67 Wh consommés`

En espérant que ce soit plus clair !

**Reply by webjib** [reply]
Ah merci pour l’explication c’est parfaitement clair !!

**Comment by bokub** [comment]
Salut @webjib !
J'ai un peu de temps libre aujourd'hui, je vais essayer de voir ce que j'arrive à faire avec les Add-ons Home Assistant (c'est du Docker, donc ça me permet d'éviter de coder en Python :slightly_smiling_face:  )

**Reply by webjib** [reply]
Moi j’ai un truc fonctionnel mais beaucoup de trucs à améliorer pour le rendre public. Mais je suis bien d’accord j’aime pas le python

**Reply by bokub** [reply]
T'es aussi parti sur un Add-On ? Ou une intégration ? Ou encore autre chose ?

**Comment by webjib** [comment]
Non suis parti sur une intégration. Pour le moment j’arrive à exposer des sensors et gérer les éventuelles erreurs, je fais aussi une requête à une heure aléatoire et limite à 3 essais dans la journée. J’ai pas encore réussi à exposer les valeurs dans le tableau de bord energy de HASS.

**Comment by bokub** [comment, marked as answer]
Salut @webjib 

Je viens de publier un nouvel add-on: [HA Linky](https://github.com/bokub/ha-linky) :partying_face: 

Je te laisse l'essayer et me dire si ça correspond à ton besoin

N'hésite pas à me faire des retours pour que je puisse l'améliorer

Bonne journée !

**Comment by webjib** [comment]
Je teste ça rapidement et te fais un retour ;-)

**Comment by webjib** [comment]
La première réponse de l’API est un 403. C’est bizarre car ma propre extension ne fonctionne plus depuis quelques jours mais je ne suis pas chez moi, c’est peut être aussi un 403. Mon IP ou mon token sont bloqués c’est possible ?

En tout cas la configuration est simple ! Par contre en amélioration ça pourrait être sympa de pouvoir suivre les couts (je ne peux pas sélectionner l’option de suivre les couts avec un prix fixe), et que certains sensor soient exposés en dehors du tableau de bord Énergie :-)

**Reply by bokub** [reply]
Est-ce que tu arrives à afficher https://conso.boris.sh depuis cette IP ?

Si oui ça veut dire que ton IP n'est pas bloquée

**Reply by webjib** [reply]
Je teste ça en rentrant chez moi alors, car j’ai installé ton add on depuis mon Raspberry à distance.

édit: j’ai lancé un Ping depuis le Raspberry et tout semble ok. Au pire je refais la procédure pour autoriser ton api auprès d’enedis ?

**Reply by webjib** [reply]
Il se peut que mon extension ait été bugguee et à lance plusieurs fetch qui ont abouti à un ban. Y’a quelques choses à faire auprès de Cloudflare?

`This website is using a security service to protect itself from online attacks. The action you just performed triggered the security solution. There are several actions that could trigger this block including submitting a certain word or phrase, a SQL command or malformed data.`
